skbeam/core/correlation.py

```diff
@@ -405,3 +407,355 @@ def auto_corr_scat_factor(lags, beta, relaxation_rate, baseline=1):
+        the current multi-tau level
+    buf_no : int
+        the current buffer number
+    """
+    img_per_level[level] += 1
+
+    # in multi-tau correlation other than first level all other levels
+    #  have to do the half of the correlation
+    if level == 0:
+        i_min = 0
+    else:
+        i_min = num_bufs//2
+
+    for i in range(i_min, min(img_per_level[level], num_bufs)):
+        t_index = level*num_bufs/2 + i
+
+        delay_no = (buf_no - i) % num_bufs
+
+        past_img = buf[level, delay_no]
+        future_img = buf[level, buf_no]
+
+        #  get the matrix of correlation function without normalizations
+        tmp_binned = (np.bincount(label_array,
+                                  weights=past_img*future_img)[1:])
+        # get the matrix of past intensity normalizations
+        pi_binned = (np.bincount(label_array,
+                                 weights=past_img)[1:])
+
+        # get the matrix of future intensity normalizations
+        fi_binned = (np.bincount(label_array,
+                                 weights=future_img)[1:])
+
+        tind1 = (current_img_time - 1)
+
+        tind2 = (current_img_time - lag_steps[t_index] - 1)
+
+        if not isinstance(current_img_time, int):
+            nshift = 2**(level-1)
+            for i in range(-nshift+1, nshift+1):
+                two_time[int(tind1+i),
+                        int(tind2+i)] = (tmp_binned/(pi_binned *
+                                                 fi_binned))*num_pixels
+        else:
+            two_time[tind1, tind2] = tmp_binned/(pi_binned *
+                                                   fi_binned)*num_pixels
+
+
+def _init_state_two_time(num_levels, num_bufs, labels, num_frames):
+    """Initialize a stateful namedtuple for two time correlation
+
+    Parameters
+    ----------
+    num_levels : int
+    num_bufs : int
+    labels : array
+        Two dimensional labeled array that contains ROI information
+
+    Returns
+    -------
+    internal_state : namedtuple
+        The namedtuple that contains all the state information that
+        `lazy_one_time` requires so that it can be used to pick up processing
+        after it was interrupted
+    """
+    (label_array, pixel_list, num_rois, num_pixels, lag_steps, buf,
+     img_per_level, track_level, cur) = _inputs(num_bufs, num_levels, labels)
+
+    # to count images in each level
+    count_level = np.zeros(num_levels, dtype=np.int64)
+    # current image time
+    #current_img_time = 0
+
+    # two time correlation results (array)
+    two_time = np.zeros((num_frames, num_frames,
+                         num_rois), dtype=np.float64)
+
+    return _two_time_internal_state(
+        buf,
+        img_per_level,
+        label_array,
+        track_level,
+        cur,
+        pixel_list,
+        num_pixels,
+        lag_steps,
+        two_time,
+        count_level,
+        #current_img_time,
+    )
+
+
+def _inputs(num_bufs, num_levels, labels):
```

Comment:
can you rename this to `_validate_and_transform_inputs` or something equally verbose